src/contracts/GPv2Settlement.sol

```diff
@@ -310,14 +310,92 @@ contract GPv2Settlement is GPv2Signing, ReentrancyGuard, StorageAccessible {
         );
     }
 
+    /// @dev Executes a single trade on-chain by performing the specified
+    /// transfers and interactions. This function is used as part of the
+    /// "fast-path" for settling single trades against on-chain liquidity.
+    ///
+    /// This method computes the executed sell amount from the total transfer
+    /// amount and the executed buy amount by reading the receiver's balance
+    /// before and after the execution.
+    ///
+    /// @param recoveredOrder The recovered order to execute.
+    /// @param transfers Direct transfers of user funds to permorm for executing
+    /// this order.
+    /// @param interactions Smart contract interaction to perform with the
+    /// transferred user funds.
+    /// @param feeDiscount The discount applied to the final executed fees.
+    function executeSingleTrade(
+        RecoveredOrder memory recoveredOrder,
+        GPv2AllowanceManager.Transfer[] calldata transfers,
+        GPv2Interaction.Data[] calldata interactions,
+        uint256 feeDiscount
+    ) internal {
+        GPv2Order.Data memory order = recoveredOrder.data;
+        bytes memory orderUid = recoveredOrder.uid;
+        uint256 executedFeeAmount = order.feeAmount.sub(feeDiscount);
+
+        uint256 executedSellAmount;
+        uint256 executedBuyAmount;
+        {
+            address receiver = recoveredOrder.receiver;
+            uint256 startingBalance = order.buyToken.balanceOf(receiver);
+
+            executedSellAmount = allowanceManager.transferToTargets(
+                order.sellToken,
+                recoveredOrder.owner,
+                transfers
+            );
+
+            executeInteractions(interactions);
+
+            executedBuyAmount = order.buyToken.balanceOf(receiver).sub(
+                startingBalance
+            );
+        }
+
+        // solhint-disable-next-line not-rely-on-time
+        require(order.validTo >= block.timestamp, "GPv2: order expired");
```

Comment:
ubernit: Is there a reason for doing the requires so late in the execution? Might be slightly more gas efficient to fail early (before the interactions) for the things we can validate early.